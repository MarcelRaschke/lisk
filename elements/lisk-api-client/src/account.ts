/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { codec } from '@liskhq/lisk-codec';
import { Channel, RegisteredSchemas } from './types';

export class Account {
	private readonly _channel: Channel;
	private readonly _schemas: RegisteredSchemas;

	public constructor(channel: Channel, schemas: RegisteredSchemas) {
		this._channel = channel;
		this._schemas = schemas;
	}

	public async get(address: Buffer | string): Promise<Record<string, unknown>> {
		const addressString: string = Buffer.isBuffer(address) ? address.toString('hex') : address;
		const accountHex: string = await this._channel.invoke('app:getAccount', {
			address: addressString,
		});

		return this.decode(Buffer.from(accountHex, 'hex'));
	}

	public encode(input: Record<string, unknown>): Buffer {
		return codec.encode(this._schemas.account, input);
	}

	public decode(input: Buffer | string): Record<string, unknown> {
		const inputBuffer: Buffer = Buffer.isBuffer(input) ? input : Buffer.from(input, 'hex');
		return codec.decode(this._schemas.account, inputBuffer);
	}

	public toJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.toJSON(this._schemas.account, account);
	}

	public fromJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.fromJSON(this._schemas.account, account);
	}
}
