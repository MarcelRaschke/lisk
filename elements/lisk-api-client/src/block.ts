/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { codec } from '@liskhq/lisk-codec';
import { Block as BlockType, Channel, RegisteredSchemas } from './types';
import { decodeBlock, encodeBlock, getTransactionAssetSchema } from './codec';

export class Block {
	private readonly _channel: Channel;
	private readonly _schemas: RegisteredSchemas;

	public constructor(channel: Channel, registeredSchema: RegisteredSchemas) {
		this._channel = channel;
		this._schemas = registeredSchema;
	}

	public async get(id: Buffer | string): Promise<Record<string, unknown>> {
		const idString: string = Buffer.isBuffer(id) ? id.toString('hex') : id;
		const blockHex = await this._channel.invoke<string>('app:getBlockByID', {
			id: idString,
		});
		const blockBytes = Buffer.from(blockHex, 'hex');
		return decodeBlock(blockBytes, this._schemas);
	}

	public async getByHeight(height: number): Promise<Record<string, unknown>> {
		const blockHex = await this._channel.invoke<string>('app:getBlockByHeight', { height });
		const blockBytes = Buffer.from(blockHex, 'hex');
		return decodeBlock(blockBytes, this._schemas);
	}

	public encode(input: {
		header: Record<string, unknown>;
		payload: Record<string, unknown>[];
	}): Buffer {
		return encodeBlock(input, this._schemas);
	}

	public decode<T = Record<string, unknown>>(input: Buffer | string): T {
		const inputBuffer: Buffer = Buffer.isBuffer(input) ? input : Buffer.from(input, 'hex');
		return decodeBlock(inputBuffer, this._schemas) as T;
	}

	public toJSON(
		block: BlockType,
	): {
		header: Record<string, unknown>;
		payload: Record<string, unknown>[];
	} {
		const { asset, ...headerRoot } = block.header;

		// We need to do this as our schemas do not include the ID. Keep this.
		const tmpBlockId = headerRoot.id;
		delete headerRoot.id;

		// decode header
		const header = {
			...codec.toJSON(this._schemas.blockHeader, headerRoot),
			asset: {},
			id: tmpBlockId?.toString('hex'),
		};

		// decode header's asset
		const headerAssetJson = codec.toJSON(
			this._schemas.blockHeadersAssets[block.header.version],
			asset,
		);
		header.asset = headerAssetJson;

		const payload: Record<string, unknown>[] = [];

		// decode transactions
		for (const tx of block.payload) {
			const { asset: txAsset, ...txRoot } = tx;
			// We need to do this as our schemas do not include the ID. Keep this.
			const tmpId = txRoot.id;
			delete txRoot.id;

			const schemaAsset = getTransactionAssetSchema(tx, this._schemas);
			const jsonTxAsset = codec.toJSON(schemaAsset, txAsset as object);
			const jsonTxRoot = codec.toJSON(this._schemas.transaction, txRoot);

			const jsonTx = {
				...jsonTxRoot,
				id: tmpId?.toString('hex'),
				asset: jsonTxAsset,
			};

			payload.push(jsonTx);
		}

		return { header, payload };
	}

	public fromJSON(
		block: BlockType<string>,
	): {
		header: Record<string, unknown>;
		payload: Record<string, unknown>[];
	} {
		const { asset, ...headerRoot } = block.header;

		// We need to do this as our schemas do not include the ID. Keep this.
		const tmpBlockId = headerRoot.id ? Buffer.from(headerRoot.id, 'hex') : Buffer.alloc(0);
		delete headerRoot.id;

		// decode header
		const header = {
			...codec.fromJSON(this._schemas.blockHeader, headerRoot),
			asset: {},
			id: tmpBlockId,
		};

		// decode header's asset
		const headerAssetJson = codec.fromJSON(
			this._schemas.blockHeadersAssets[block.header.version],
			asset,
		);
		header.asset = headerAssetJson;

		const payload: Record<string, unknown>[] = [];
		// decode transactions
		for (const tx of block.payload) {
			const { asset: txAsset, ...txRoot } = tx;
			// We need to do this as our schemas do not include the ID. Keep this.
			const tmpId = txRoot.id ? Buffer.from(txRoot.id, 'hex') : Buffer.alloc(0);
			delete txRoot.id;

			const schemaAsset = getTransactionAssetSchema(tx, this._schemas);
			const txAssetObject = codec.fromJSON(schemaAsset, txAsset as object);
			const txRootObject = codec.fromJSON(this._schemas.transaction, txRoot);

			const txObject = {
				...txRootObject,
				id: tmpId,
				asset: txAssetObject,
			};

			payload.push(txObject);
		}

		return { header, payload };
	}
}
