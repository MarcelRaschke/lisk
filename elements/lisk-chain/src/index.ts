/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK, EVENT_VALIDATORS_CHANGED } from './constants';

const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK, EVENT_VALIDATORS_CHANGED };

export { events };
export { Chain } from './chain';
export { Transaction, TransactionInput, transactionSchema, calculateMinFee } from './transaction';
export {
	blockHeaderAssetSchema,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	validatorsSchema,
	getGenesisBlockHeaderAssetSchema,
	stateDiffSchema,
	getRegisteredBlockAssetSchema,
} from './schema';
export {
	CONSENSUS_STATE_VALIDATORS_KEY,
	CONSENSUS_STATE_FINALIZED_HEIGHT_KEY,
	TAG_BLOCK_HEADER,
	TAG_TRANSACTION,
} from './constants';
export type {
	Account,
	AccountDefaultProps,
	RawBlock,
	RawBlockHeader,
	GenesisBlock,
	GenesisBlockHeader,
	Block,
	BlockHeader,
	BlockHeaderAsset,
	Validator,
	AccountSchema,
} from './types';
export { Slots } from './slots';
export { readGenesisBlockJSON, getValidators, getAccountSchemaWithDefault } from './utils';
export * as testing from './testing';

export type { StateStore } from './state_store';
export type { DataAccess } from './data_access';
