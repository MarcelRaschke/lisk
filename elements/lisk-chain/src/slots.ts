/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

interface SlotsInput {
	readonly genesisBlockTimestamp: number;
	readonly interval: number;
}

const SEC_IN_MS = 1000;

export class Slots {
	private readonly _genesisTime: number;
	private readonly _interval: number;

	public constructor({ genesisBlockTimestamp, interval }: SlotsInput) {
		this._genesisTime = genesisBlockTimestamp;
		this._interval = interval;
	}

	public timeSinceGenesis(): number {
		return Math.floor((Date.now() - this._genesisTime * SEC_IN_MS) / SEC_IN_MS);
	}

	public blockTime(): number {
		return this._interval;
	}

	public getSlotNumber(timeStamp?: number): number {
		const time = timeStamp ?? Math.floor(Date.now() / SEC_IN_MS);
		const elapsedTime = time - this._genesisTime;

		return Math.floor(elapsedTime / this._interval);
	}

	public getSlotTime(slot: number): number {
		const slotGenesisTimeOffset = slot * this._interval;

		return this._genesisTime + slotGenesisTimeOffset;
	}

	public getNextSlot(): number {
		const slot = this.getSlotNumber();

		return slot + 1;
	}

	public isWithinTimeslot(slot: number, time?: number): boolean {
		return this.getSlotNumber(time) === slot;
	}
}
