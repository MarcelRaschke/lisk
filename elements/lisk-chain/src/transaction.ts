/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { hash, getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
import { validator, LiskValidationError } from '@liskhq/lisk-validator';

export interface TransactionInput {
	readonly moduleID: number;
	readonly assetID: number;
	readonly senderPublicKey: Buffer;
	readonly nonce: bigint;
	readonly fee: bigint;
	readonly asset: Buffer;
	readonly signatures: ReadonlyArray<Buffer>;
}

export const transactionSchema = {
	$id: 'lisk/transaction',
	type: 'object',
	required: ['moduleID', 'assetID', 'nonce', 'fee', 'senderPublicKey', 'asset'],
	properties: {
		moduleID: {
			dataType: 'uint32',
			fieldNumber: 1,
			minimum: 2,
		},
		assetID: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		nonce: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		fee: {
			dataType: 'uint64',
			fieldNumber: 4,
		},
		senderPublicKey: {
			dataType: 'bytes',
			fieldNumber: 5,
			minLength: 32,
			maxLength: 32,
		},
		asset: {
			dataType: 'bytes',
			fieldNumber: 6,
		},
		signatures: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 7,
		},
	},
};

export const calculateMinFee = (
	tx: Transaction,
	minFeePerByte: number,
	baseFees: { moduleID: number; assetID: number; baseFee: string }[],
): bigint => {
	const size = tx.getBytes().length;
	const baseFee =
		baseFees.find(bf => bf.moduleID === tx.moduleID && bf.assetID === tx.assetID)?.baseFee ?? '0';
	return BigInt(minFeePerByte * size) + BigInt(baseFee);
};

export class Transaction {
	public readonly moduleID: number;
	public readonly assetID: number;
	public readonly asset: Buffer;
	public readonly nonce: bigint;
	public readonly fee: bigint;
	public readonly senderPublicKey: Buffer;
	public readonly signatures: ReadonlyArray<Buffer>;
	private _id?: Buffer;
	private _senderAddress?: Buffer;

	public constructor(transaction: TransactionInput) {
		this.moduleID = transaction.moduleID;
		this.assetID = transaction.assetID;
		this.asset = transaction.asset;
		this.nonce = transaction.nonce;
		this.fee = transaction.fee;
		this.senderPublicKey = transaction.senderPublicKey;
		this.signatures = transaction.signatures;
	}

	public static decode(bytes: Buffer): Transaction {
		const tx = codec.decode<TransactionInput>(transactionSchema, bytes);
		return new Transaction(tx);
	}

	public get id(): Buffer {
		if (!this._id) {
			this._id = hash(this.getBytes());
		}
		return this._id;
	}

	public get senderAddress(): Buffer {
		if (!this._senderAddress) {
			this._senderAddress = getAddressFromPublicKey(this.senderPublicKey);
		}
		return this._senderAddress;
	}

	public getBytes(): Buffer {
		const transactionBytes = codec.encode(transactionSchema, this as Record<string, unknown>);

		return transactionBytes;
	}

	public getSigningBytes(): Buffer {
		const transactionBytes = codec.encode(transactionSchema, ({
			...this,
			signatures: [],
		} as unknown) as Record<string, unknown>);

		return transactionBytes;
	}

	public validate(input: {
		minFeePerByte: number;
		baseFees: { moduleID: number; assetID: number; baseFee: string }[];
	}): void {
		const schemaErrors = validator.validate(transactionSchema, this);
		if (schemaErrors.length > 0) {
			throw new LiskValidationError(schemaErrors);
		}
		if (this.signatures.length === 0) {
			throw new Error('Signatures must not be empty');
		}
		for (const signature of this.signatures) {
			if (signature.length !== 0 && signature.length !== 64) {
				throw new Error('Signature must be empty or 64 bytes');
			}
		}
		const minFee = calculateMinFee(this, input.minFeePerByte, input.baseFees);
		if (this.fee < minFee) {
			throw new Error(
				`Insufficient transaction fee. Minimum required fee is: ${minFee.toString()}`,
			);
		}
	}
}
