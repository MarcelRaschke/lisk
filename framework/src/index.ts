/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export {
	Transaction,
	transactionSchema,
	readGenesisBlockJSON,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	getAccountSchemaWithDefault,
	getGenesisBlockHeaderAssetSchema,
	blockHeaderAssetSchema,
} from '@liskhq/lisk-chain';
export {
	BaseModule,
	BaseAsset,
	TokenModule,
	TokenTransferAsset,
	SequenceModule,
	KeysModule,
	KeysRegisterAsset,
	DPoSModule,
	DPoSRegisterAsset,
	DPoSVoteAsset,
	DPoSUnlockAsset,
	DPoSPoMAsset,
} from './modules';
export { Application } from './application';
export { systemDirs } from './system_dirs';
export { BasePlugin, BlockHeaderJSON, PluginInfo, PluginCodec } from './plugins/base_plugin';
export { IPCChannel } from './controller/channels';
export type { BaseChannel } from './controller/channels';
export type { EventsDefinition, EventCallback } from './controller/event';
export type { ActionsDefinition, ActionHandler } from './controller/action';
export * as testing from './testing';
export * from './types';
