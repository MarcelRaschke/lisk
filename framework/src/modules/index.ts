/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export * from './base_module';
export * from './base_asset';
export { KeysModule, RegisterAsset as KeysRegisterAsset } from './keys';
export { TokenModule, TransferAsset as TokenTransferAsset } from './token';
export { SequenceModule } from './sequence';
export {
	DPoSModule,
	RegisterTransactionAsset as DPoSRegisterAsset,
	VoteTransactionAsset as DPoSVoteAsset,
	UnlockTransactionAsset as DPoSUnlockAsset,
	PomTransactionAsset as DPoSPoMAsset,
} from './dpos';
