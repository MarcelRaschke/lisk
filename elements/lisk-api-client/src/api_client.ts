/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { EventCallback, Channel, RegisteredSchemas, NodeInfo } from './types';
import { Node } from './node';
import { Account } from './account';
import { Block } from './block';
import { Transaction } from './transaction';

export class APIClient {
	private readonly _channel: Channel;
	private _schemas!: RegisteredSchemas;
	private _nodeInfo!: NodeInfo;
	private _node!: Node;
	private _account!: Account;
	private _block!: Block;
	private _transaction!: Transaction;

	public constructor(channel: Channel) {
		this._channel = channel;
	}

	public async init(): Promise<void> {
		this._schemas = await this._channel.invoke<RegisteredSchemas>('app:getSchema');
		this._node = new Node(this._channel);
		this._account = new Account(this._channel, this._schemas);
		this._block = new Block(this._channel, this._schemas);
		this._nodeInfo = await this._node.getNodeInfo();
		this._transaction = new Transaction(this._channel, this._schemas, this._nodeInfo);
	}

	public async disconnect(): Promise<void> {
		return this._channel.disconnect();
	}

	public async invoke<T = Record<string, unknown>>(
		actionName: string,
		params?: Record<string, unknown>,
	): Promise<T> {
		return this._channel.invoke(actionName, params);
	}

	public subscribe(eventName: string, cb: EventCallback): void {
		this._channel.subscribe(eventName, cb);
	}

	public get schemas(): RegisteredSchemas {
		return this._schemas;
	}

	public get node(): Node {
		return this._node;
	}

	public get account(): Account {
		return this._account;
	}

	public get block(): Block {
		return this._block;
	}

	public get transaction(): Transaction {
		return this._transaction;
	}
}
