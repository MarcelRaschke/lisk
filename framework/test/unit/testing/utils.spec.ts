/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { APIClient } from '@liskhq/lisk-api-client';
import { SequenceModule, TokenModule } from '../../../src';
import {
	getAccountSchemaFromModules,
	getModuleInstance,
	waitUntilBlockHeight,
} from '../../../src/testing/utils';

describe('utils', () => {
	describe('getAccountSchemaFromModules', () => {
		it('should get schema object for modules', () => {
			expect(getAccountSchemaFromModules([TokenModule, SequenceModule])).toMatchSnapshot();
		});
	});

	describe('getModuleInstance', () => {
		it('should create module instance with default mocks', () => {
			const module = getModuleInstance(TokenModule);

			expect(module).toBeInstanceOf(TokenModule);
		});

		it('should create module instance with custom data access', () => {
			const dataAccess = {
				getChainState: jest.fn(),
				getAccountByAddress: jest.fn(),
				getLastBlockHeader: jest.fn(),
			};
			const module = getModuleInstance(TokenModule, { dataAccess });

			expect(module).toBeInstanceOf(TokenModule);
			expect(module['_dataAccess']).toBe(dataAccess);
		});

		it('should create module instance with custom logger', () => {
			const logger = {
				trace: jest.fn(),
				debug: jest.fn(),
				info: jest.fn(),
				warn: jest.fn(),
				error: jest.fn(),
				fatal: jest.fn(),
				level: jest.fn(),
			};
			const module = getModuleInstance(TokenModule, { logger });

			expect(module).toBeInstanceOf(TokenModule);
			expect(module['_logger']).toBe(logger);
		});

		it('should create module instance with custom channel', () => {
			const channel = {
				publish: jest.fn(),
			};
			const module = getModuleInstance(TokenModule, { channel });

			expect(module).toBeInstanceOf(TokenModule);
			expect(module['_channel']).toBe(channel);
		});
	});

	describe('waitUntilBlockHeight', () => {
		let apiClient: APIClient;
		beforeEach(() => {
			apiClient = {
				// eslint-disable-next-line @typescript-eslint/require-await
				subscribe: jest.fn(async (_, callback) => callback({ block: 'blockdata' })),
				block: {
					decode: jest.fn().mockReturnValue({ header: { height: 2 } }),
				},
			} as any;
		});

		it('should resolve after input height', async () => {
			await expect(waitUntilBlockHeight({ apiClient, height: 1 })).resolves.toBeUndefined();
		});

		it('should timeout', async () => {
			// eslint-disable-next-line @typescript-eslint/require-await
			apiClient.subscribe = jest.fn(async () => setTimeout(() => undefined, 2));
			await expect(waitUntilBlockHeight({ apiClient, height: 1, timeout: 1 })).rejects.toThrow(
				"'waitUntilBlockHeight' timed out after 1 ms",
			);
		});
	});
});
