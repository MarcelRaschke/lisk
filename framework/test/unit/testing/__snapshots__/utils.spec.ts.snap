// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getAccountSchemaFromModules should get schema object for modules 1`] = `
Object {
  "sequence": Object {
    "default": Object {
      "nonce": 0n,
    },
    "fieldNumber": 3,
    "properties": Object {
      "nonce": Object {
        "dataType": "uint64",
        "fieldNumber": 1,
      },
    },
    "type": "object",
  },
  "token": Object {
    "default": Object {
      "balance": 0n,
    },
    "fieldNumber": 2,
    "properties": Object {
      "balance": Object {
        "dataType": "uint64",
        "fieldNumber": 1,
      },
    },
    "type": "object",
  },
}
`;
